STYLEGUIDE.adoc

```diff
@@ -0,0 +1,426 @@
+
+==== Examples
+Bad:
+```
+data Foo = Foo Int
+```
+Good:
+```
+data Foo = MkFoo Int
+```
+
+=== Avoid ticks in function names
+Generally avoid using ticks to distinguish function names. All this conveys is that it is “another” version of the function. Try expressing the difference in the function name, even if it makes it longer.
+
+==== Examples
+
+Bad:
+```
+runCek
+runCek’
+```
+Good:
+```
+runCek
+runCekWithLogs
+```
+
+==== Reasoning
+
+It’s a tempting naming convention, but no one likes reading code with such functions. The function names should convey helpful information when possible.
+
+=== Avoid multi-line parenthesized expressions
+
+A parenthesized expression should not span multiple lines. Pull it out to a named binding, use `$`, or otherwise reorganize the code.
+
+==== Examples
+
+Bad:
+```haskell
+foldr (\a acc -> let
+    x = ...
+  in a + x + acc) x xs
+```
+Good:
+```
+foldr meaningfulName x xs
+  where
+    meaningfulName :: ...
+    meaningfulName a acc = let x = ... in a + x + acc
+```
+
+==== Reasoning
+
+A parenthesis forces the user to keep a stack in their head to remember when the current "argument" finishes. Line length limits this to some degree, but if we allow line breaks then the amount of stack can become quite unwieldy.
+
+This also explains why `$` is good: since it indicates there will be no closing paren, there is no need for a stack (it's the "tail call" of bracketing).
+
+=== Use deriving strategies
+
+Always use deriving strategies.
+
+==== Reasoning
+Not using deriving strategies requires the compiler to guess which strategy you want. This can have consequences, especially when `DeriveAnyClass` is enabled, since you can accidentally end up using anyclass deriving when you didn’t mean to. Better to be explicit.
+
+=== Use Text
+
+Use Text instead of String unless you have a good reason not to.
+
+==== Reasoning
+
+It’s 2022, use a proper, unicode aware string type instead of a linked list.
+
+=== Use the package name as the root of the module name hierarchy
+
+If the package is `foo-bar`, then the modules should all be `FooBar.X`.
+
+==== Reasoning
+
+See “Naming conventions” https://www.haskellforall.com/2021/05/module-organization-guidelines-for.html[here]. We do it slightly differently (“FooBar” rather than “Foo.Bar”), but the main principle is the same.
+
+=== Reading direction
+
+Try to keep a single line *mostly* reading left-to-right or right-to-left.
+
+==== Examples
+
+Bad:
+```
+traverse (\x -> <some long function body) things
+```
+Good:
+```
+for xs $ \x -> …
+```
+
+==== Reasoning
+
+Haskell can get quite condensed and hard to read, especially when the reading direction changes frequently. Often there are symmetrical versions of operators like `<=<` and `>=>` or `=<<` and `>>=` that you can switch between to make code easier to read.
+=== When to use hierarchical re-exporting modules
+
+TODO
```

Comment:
I don't generally think it's worth including such TODOs, I'm not sure it makes us any more likely to remember to write something.